### PR TITLE
GLM COEFFICIENTS FIX

### PR DESCRIPTION
**Pull request by koniecsveta**
https://0xdata.atlassian.net/browse/PUBDEV-7118
With this change, glm model with standardize=False will calculate also the standardizedCoefficients (as in standardize=True case).
Note to be checked: it stands in documentation (model_base.py), that if standardize=False then coef() return the coefficients which are fit directly - does this mean that varimp should be calculated from these and not from newly calculated standardized ones?

### CODE REVIEW

**Comment by wendycwong** [comment]
Zuzana:  

Can you add the JIRA link to the beginning of your PR please?

Regarding your question, regardless of what value the user sets to the parameter standardize, the user can always call model.coef_norm() in Python/R to get the standardized coefficient.  In Java, you wrote the other routine for that.  Hence, if anyone wants the coefficient to get an idea of varimp, they should always call coef_norm().

File: h2o-algos/src/main/java/hex/DataInfo.java [hunk 1]
```diff
@@ -377,6 +379,28 @@ public DataInfo validDinfo(Frame valid) {
     }
     return beta;
   }
+  
+  public double[] denormalizeBeta(double [] beta, boolean standardize){
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/DataInfo.java]: Zuzana:  It does not look like we need this function if standardize=false, right?

Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/DataInfo.java]: no I think no, pushed the normalizeBeta(double [] beta, boolean standardize) and left this one for now to be able to switch fast during tests. In the end I will remove it.

File: h2o-algos/src/test/java/hex/glm/GLMTest.java [hunk 2]
```diff
@@ -109,6 +109,78 @@ private void checkScore(long rid, double [] predictions, double [] outputs){
...
+      }
+      new TestUtil.StandardizeColumns(numCols2Transform, colMeans, oneOSigma, train).doAll(train);
+      DKV.put(train);
+      Scope.track(train);
+      
+      params._standardize=false;
+      params._train = train._key;
+      GLMModel glmS = new GLM(params).trainModel().get();
+      Scope.track_generic(glmS);
+
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMTest.java]: I have added one more test.  Can you replace line 153 to line line 158 with the following code: 

      HashMap<String, Double> cMeans = new HashMap<>();
      HashMap<String, Double> cSigmas = new HashMap<>();
      String[] cnames = train.names();
      for (int cindex = numStart; cindex < response_index; cindex++) {
        numCols2Transform[countIndex]=cindex;
        colMeans[countIndex] = train.vec(cindex).mean();
        oneOSigma[countIndex] = 1.0/train.vec(cindex).sigma();
        cMeans.put(cnames[cindex], colMeans[countIndex]);
        cSigmas.put(cnames[cindex], train.vec(cindex).sigma());
        countIndex++;
      }
      
      params._standardize = false;  // build a model on non-standardized columns with no standardization.
      GLMModel glmF = new GLM(params).trainModel().get();
      Scope.track_generic(glmF);

      HashMap<String, Double> coeffSF = glmF.standardizedCoefficients();
      HashMap<String, Double> coeffF = glmF.coefficients();
      if (family.equals(Family.multinomial)) {
        double[] interPClass = new double[glmF._output._nclasses];
        for (String key : coeffSF.keySet()) {
          double temp1 = coeffSF.get(key);
          double temp2 = coeffF.get(key);
          if (Math.abs(temp1 - temp2) > 1e-6) { // coefficient same for categoricals, different for numericals
            String[] coNames = key.split("_");
            if (!(coNames[1].equals("Intercept"))) {  // skip over intercepts
              String colnames = coNames[1];
              interPClass[Integer.valueOf(coNames[0])] += temp2 * cMeans.get(colnames);
              temp2 = temp2 * cSigmas.get(colnames);
            }
            assert Math.abs(temp1 - temp2) < 1e-6 : "Expected coefficients for " + coNames[1] + " is " + temp1 + " but actual " + temp2;
          }
        }
        // check for equality of intercepts
        for (int index = 0; index < glmF._output._nclasses; index++) {
          String interceptKey = index + "_Intercept";
          double temp1 = coeffSF.get(interceptKey) + interPClass[index];
          double temp2 = coeffF.get(interceptKey);
          assert Math.abs(temp1 - temp2) < 1e-6 : "Expected coefficients for " + interceptKey + " is " + temp1 + " but actual "
                  + temp2;
        }
      } else {
        double interceptOffset = 0;
        for (String key:coeffF.keySet()) {
          double temp1 = coeffSF.get(key);
          double temp2 = coeffF.get(key);
          if (Math.abs(temp1 - temp2) > 1e-6) {
            if (!key.equals("Intercept")) {
              interceptOffset += temp2*cMeans.get(key);
              temp2 = temp2*cSigmas.get(key);
              assert Math.abs(temp1 - temp2) < 1e-6 : "Expected coefficients for " + key + " is " + temp1 + " but actual " + temp2;
            }
          }
        }
        // check intercept terms 
        double temp1 = coeffSF.get("Intercept");
        double temp2 = coeffF.get("Intercept")+interceptOffset;
        assert Math.abs(temp1 - temp2) < 1e-6 : "Expected coefficients for Intercept is " + temp1 + " but actual "
                + temp2;
      }

File: h2o-r/h2o-package/R/models.R [hunk 3]
```diff
@@ -1469,6 +1469,28 @@ h2o.coef_norm <- function(object) {
   }
 }
 
+grabCoeff <- function(tempTable, nameStart, standardize=FALSE) {
+    coeffNamesPerClass <- tempTable$names # contains coeff names per class
+    totTableLength <- length(tempTable)
+    numClass <- totTableLength
+    startIndex <- 2
+    endIndex <- numClass
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/models.R]: please delete line 1475 numClass <- totTableLength

Please replace line 1477 endIndex <- numClass with the following line:
    endIndex <- (totTableLength-1)/2+1

File: h2o-algos/src/test/java/hex/glm/GLMTest.java [hunk 4]
```diff
@@ -109,6 +109,132 @@ private void checkScore(long rid, double [] predictions, double [] outputs){
...
+          double temp1 = coeffSF.get(key);
+          double temp2 = coeffF.get(key);
+          if (Math.abs(temp1 - temp2) > 1e-6) { // coefficient same for categoricals, different for numericals
+            String[] coNames = key.split("_");
+            if (!(coNames[1].equals("Intercept"))) {  // skip over intercepts
+              String colnames = coNames[1];
+              interPClass[Integer.valueOf(coNames[0])] += temp2 * cMeans.get(colnames);
+              temp2 = temp2 * cSigmas.get(colnames);
+            }
+            assert Math.abs(temp1 - temp2) < 1e-6 : "Expected coefficients for " + coNames[1] + " is " + temp1 + " but actual " + temp2;
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMTest.java]: Can you exchange the lines 182 and l83.  It attempts to compare intercept before it is ready

File: h2o-algos/src/test/java/hex/glm/GLMTest.java [hunk 5]
```diff
@@ -109,6 +109,132 @@ private void checkScore(long rid, double [] predictions, double [] outputs){
...
+          }
+        }
+        // check for equality of intercepts
+        for (int index = 0; index < glmF._output._nclasses; index++) {
+          String interceptKey = index + "_Intercept";
+          double temp1 = coeffSF.get(interceptKey) + interPClass[index];
+          double temp2 = coeffF.get(interceptKey);
+          assert Math.abs(temp1 - temp2) < 1e-6 : "Expected coefficients for " + interceptKey + " is " + temp1 + " but actual "
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMTest.java]: replace 189 with double temp1 = coeffSF.get(interceptKey)
line 190 with double temp2 = coeffF.get(interceptKey)+ + interPClass[index];

**Pull request merged**